Restoring specification text for `EXT_structural_metadata`

javagl opened the pull request:
The `EXT_mesh_features` extension is revised, and the part that defines property tables and property textures is moved into its own `EXT_structured_metadata` extension. This PR adds parts of the original `EXT_mesh_features` README (as of https://github.com/CesiumGS/glTF/tree/e8b8a1065ac9e3db571fa8af836fac5480020228/extensions/2.0/Vendor/EXT_mesh_features ) back into the revised state. (The part that discusses Feature IDs has been restored in https://github.com/CesiumGS/glTF/pull/51 )

Preview: https://github.com/javagl/glTF/tree/ext-mesh-features-structural-spec/extensions/2.0/Vendor/EXT_structural_metadata 

---

The adjustments until now (*updated*)

- Replaced revision history with link to common revision history for 3D Tiles Next
- Omitted the part that defined feature IDs, adjusted introduction and wording to not focus on feature IDs, but rather on the core elements (Schema definition and metadata storage)
- Added section about Property Mappings

- <strike>Verify: Added example showing how the property texture channels can be combined into other types.</strike> (Added spec text, minor schema update)
- <strike>Update/restore images and examples</strike>
- <strike>Update schema to say that min/max are only applicable for numeric component types</strike>
- <strike>... the name should probably `propertyMappings` (plural)</strike> Done via https://github.com/CesiumGS/glTF/pull/53
- <strike>Consider writing something about the "overriding" mechanism for `offset`/`scale`...</strike> Done via https://github.com/CesiumGS/glTF/pull/52/commits/86424b9a05bb9be3856b7f4c40ffa228470ba121

javagl commented:
I'll mark it as "ready for review" for now. There's always room for improvements, but it should not remain in Draft state forever. The TODOs from the comment above might be addressed during the review. Beyond that, a nice example *image* for property tables would be nice. It could be one from the [original examples](https://github.com/CesiumGS/glTF/blob/proposal-EXT_mesh_features/extensions/2.0/Vendor/EXT_mesh_features/README.md#examples) list, if the "Feature ID" part could be handled nicely. Maybe just say something like

> The row indices for accessing the table can be obtained in various ways, for example, [link to feature ID definition]

? I'm open for suggestions.

lilleyse reviewed (commented):
> I'll mark it as "ready for review" for now. There's always room for improvements, but it should not remain in Draft state forever

Agreed, certainly some sections could have more details but I think this strikes a good balance.

> Beyond that, a nice example image for property tables would be nice.

The first example of the house might be a good one (plus we have a redone SVG version of that now).

javagl commented on extensions/2.0/Vendor/EXT_structural_metadata/README.md:
```diff
@@ -11,40 +13,490 @@
+>     "EXT_structural_metadata": {
+>       "schema": { ... },
+>       "propertyTables": [{
+>         "name": "tree_survey_2021-09-29",
```
Why was it removed?

These are "complex" objects that store application-specific data. I could imagine that there will sometimes be the necessity to present _some_ information about them in the UI: "This table stores the data that was obtained via survey 2021-09-29", vs. "This one is for 2021-09-30" ...

(The question is remotely related to the questions about "tags" for "Feature ID sets": How will the user select the data that is visualized, and how will he _know_ what this data _is_?).

I think that at least having the _option_ to transport some "application/domain-specific information" there could be desirable. This could be a plain `name`, or a `description`. If it is only a name, this could, by the way, also be accomplished by not extending `glTFProperty`, but [`glTFChildOfRootProperty`](https://github.com/javagl/glTF/blob/60efccc60e05e5764a48722a37d4a0a5fae46aac/specification/2.0/schema/glTFChildOfRootProperty.schema.json).

javagl replied:
(Unless that "...ChildOfRoot..." is not considered to be not-fitting, because it's rather a "...ChildOfExtension..." :-/ )

lilleyse replied:
Those are good points, some application/domain-specific information is helpful. I'll add `name` back.

lilleyse replied:
My original thinking was that a feature ID set plus class name and description was sufficient to describe the data, but now I agree that property table `name` is useful too.

lilleyse replied:
Pushed the schema changes:

https://github.com/CesiumGS/glTF/commit/9486ece8a3d33bf243f0c06c0cce9718a574b14c
https://github.com/CesiumGS/3d-tiles/commit/1746ca99df2a50a809a984f917fd8a7011579354

javagl commented on extensions/2.0/Vendor/EXT_structural_metadata/README.md:
```diff
@@ -11,40 +13,490 @@
+>   ]
+> }
+
+Multiple channels can be used to represent individual bytes of larger data types: The values from the selected channels are treated as unsigned 8 bit integers, and represent the bytes of the actual property value, in little-endian order.
```
I thought that the latest state was, roughly summarized:

- For Feature ID textures, it should be possible to assemble larger types from multiple `channels` (e.g. to have 16bit IDs)
- For "consistency", the same could/should be possible for property textures

Maybe I misunderstood this?

(Also: I'm not sure in how far the schema has to be updated. Right now, it just has a `channels: [0...]` array...)

lilleyse replied:
> For "consistency", the same could/should be possible for property textures

I thought we decided to hold off on this change to property textures, but now that I see the change I think it's a good one and we should include it.

As for schema updates, I should have been clearer. The type definition doesn't need to change but the description should include more details about how property values are packed into texture channels, similar to [the channels description for featureIdTexture.schema.json](https://github.com/CesiumGS/glTF/blob/ext-mesh-features-revision/extensions/2.0/Vendor/EXT_mesh_features/schema/featureIdTexture.schema.json#L19).

javagl commented on extensions/2.0/Vendor/EXT_structural_metadata/README.md:
```diff
@@ -11,40 +13,490 @@
+
+> **Implementation note:** Authoring tools may choose to align all buffer views to 8-byte boundaries for consistency, but client implementations should only depend on 8-byte alignment for buffer views containing 64-bit component types.
+
+### Property Mappings
```
"Property Attributes", maybe?

lilleyse replied:
Yeah let's go with that. It's more symmetric with "Property Table" and "Property Texture". The downside of course is that "Property" and "Attribute" are both somewhat general terms, but so is "Mapping".

lilleyse replied:
Is the name change still planned for this PR? Or does it require more discussion?

javagl commented:
Regarding the examples:

Originally, there have been the _inlined_ examples, marked with **"Example**", always referring to the _particular_ part of the JSON schema that the section referred to.

Independent of that there had been the "Examples" section. 

These examples had been largely independent: The inlined ones mainly contained JSON. The ones in the section only contained images and a short text. And the latter had largely been "use cases", showing different ways of how to combine Feature IDs and Property data. 

In view of the narrower focus of the new spec text, it probably makes more sense to make the inlined examples more complete, and omit the "Examples" section. 

Part of making the inlined examples more complete will be to make the JSON consistent with the images. 

(I was about to draw images of trees with 42 birds in them, to have a sensible `birdCount`. But I'll rather change the `birdCount` to an `age` or so...)

javagl commented:
- Updated the example for the property tables
- Updated the schema `description` to allow multiple channels in property textures to be combined. (I had added the [example at the bottom of this section](https://github.com/javagl/glTF/tree/ext-mesh-features-structural-spec/extensions/2.0/Vendor/EXT_structural_metadata#property-textures) a while ago, and wasn't entirely sure whether this is a desired way of describing it... I hope it's OK...)
- Added a short hint about the possibility to override `minimum/maximum/offset/scale` in the property storage section

lilleyse commented on extensions/2.0/Vendor/EXT_structural_metadata/README.md:
```diff
@@ -11,40 +13,522 @@
+Several constraints are imposed to ensure compatibility of texture storage with various property types:
+
+* Variable-length arrays are not supported.
+* Data type and bit depth of the image must be compatible with the property type.
```
This note and the note below about floating-point KTX2 might be obsolete now that properties values can be packed in multiple chanels.

lilleyse commented on extensions/2.0/Vendor/EXT_structural_metadata/README.md:
```diff
@@ -11,40 +13,522 @@
+>           "wall": {
+>             "name": "Wall Temperature vs. Insulation",
+>             "properties": {
+>               "insideTemperature": {
+>                 "name": "Inside Temperature",
+>                 "componentType": "UINT8"
+>               },
+>               "outsideTemperature": {
+>                 "name": "Outside Temperature",
+>                 "componentType": "UINT8"
+>               },
+>               "insulation": {
+>                 "name": "Insulation Thickness",
+>                 "componentType": "UINT8",
+>                 "normalized": true
+>               }
```
Add `"type": "SCALAR"` to each property definition. Looks like I missed this in my first review.

lilleyse commented on extensions/2.0/Vendor/EXT_structural_metadata/README.md:
````diff
@@ -11,40 +13,522 @@
+> }
+> ```
+
+A `propertyTexture` object extends the glTF [`textureInfo`](../../../../specification/2.0/schema/textureInfo.schema.json) object. `texCoord` specifies a texture coordinate set in the referring primitive.
````
This behavior changed between `EXT_mesh_features` and `EXT_structural_metadata`. Now each individual property extends the `textureInfo` object. This way the number of class properties is no longer constrained by the number of texture channels. The new structure looks like this:

```json
      "propertyTextures": [
        {
          "class": "wall",
          "properties": {
            "insideTemperature": {
              "index": 0,
              "texCoord": 0,
              "channels": [0]
            },
            "outsideTemperature": {
              "index": 0,
              "texCoord": 0,
              "channels": [1]
            },
            "insulation": {
              "index": 0,
              "texCoord": 0,
              "channels": [2]
            }
          }
        }
      ]
```

lilleyse commented on extensions/2.0/Vendor/EXT_structural_metadata/README.md:
````diff
@@ -11,40 +13,522 @@
+> value[1] = byte2 | (byte3 << 8);
+> ```
+
+Texture filtering must be `9728` (NEAREST), `9729` (LINEAR), or undefined, for any texture object referenced as a property texture.
````
After struggling with https://github.com/CesiumGS/cesium/pull/10111 I added a note in `EXT_mesh_features` that texture values must be encoded in linear space. Something similar should be added here.

https://github.com/CesiumGS/glTF/commit/0db466b9d32ab458affbe34379959489a46921a5

javagl replied:
Above, at https://github.com/CesiumGS/glTF/pull/52/files/399f988f41d8a9d2160d3c219088523ec90f483c#diff-7a962affc914907292f8f9f93402e72b6b49599a6bb34b57592955050339c7a1R463 , it already says 

> All values are stored in linear space.

but it may be worth mentioning here again.

lilleyse replied:
Ah, I might have just missed that. No need to mention it in two places. Though could you update the wording from `All values are stored in linear space` to `Texture values must be encoded with a linear transfer function.`

lilleyse commented on extensions/2.0/Vendor/EXT_structural_metadata/README.md:
```diff
@@ -11,40 +13,522 @@
+>     "EXT_structural_metadata": {
+>       "schema": {  
+>         "classes": {
+>           "tree": {
+>             "name": "movement",
```
```suggestion
>           "movement": {
```

javagl commented:
- Renamed "Property Mapping" to "Property Attribute"
- Integrated the smaller fixes that had been inlined (mainly smaller updates of the inlined samples)
- Moved parts of the text into a "Property Texture Data Storage" section

The latter is an attempt of a small cleanup and describe "how multiple channels are composed into larger types" in *one* dedicated place (also setting the stage for the "Example"). It now explicitly says that this is based on the assumption that there are 4 channels, and that one channel = one `byte`. The case for other formats is mentioned as an "Implementation Note" (could also just be a "Note"...), to make clear that this is not covered by this extension. 

(The inlined example that uses the `class: "wind"` could be extended, to be an example that defines two properties in _different_ textures, i.e. to use types that are so large that they need `{ index: 0, channels: [0,1] }` and `{ index: 1, channels: [0,1,2,3] }`, just to "show off". It's a trade-off between "keeping things simple" and "showing what's possible", but sooner or later we should have such an example for testing...)

lilleyse commented:
Latest version looks good. Thanks @javagl.
